proposals/0000-no-kind-equalities.rst

```diff
@@ -0,0 +1,154 @@
+so. No one asked for it. It just seemed like an easy addition. However, it has
+proved anything but easy. (This is not the first time we've been stymied supporting it.)
+So let's just send it back to the void whence it came.
+
+Ryan Scott's `blog post <https://ryanglscott.github.io/2021/08/01/equality-constraints-in-kinds/>`_
+provides additional background and gotchas around the existing feature. He even has to
+say "They're not totally useless" to defend the feature.
+
+Proposed Change Specification
+-----------------------------
+
+Disallow the constraint arrow ``=>`` in the following contexts:
+
+* A kind annotation within a type: ``f ::  Int -> (Bool :: <<here>>) -> Double``
+* A visible kind application within a type: ``f :: Proxy @(<<here>>) True -> ...``
+* A kind annotation on a type variable binder: ``f :: forall (a :: <<here>>). a -> a``
+* A kind annotation on a ``data`` declaration: ``data T a b :: <<here>> where ...``
+* A kind annotation on a family declaration: ``type family F a b :: <<here>>`` (or ``data family``)
+* A standalone kind signature: ``type T :: <<here>>``
+
+In addition, any data constructor whose type includes a constraint is
+not available for use in types.
+
+Note that class constraints are already forbidden in exactly these places,
+so this proposal changes only the status of equality constraints (both
+homogeneous, with ``~``, and heterogeneous, with ``~~``).
+
+Examples
+--------
+In addition to the example in the introduction, which would now be rejected,
+this proposal stops the use of any constrained data constructor in a type.
+For example::
+
+  data S a where
+    MkS :: F a ~ Int => a -> a -> S a
+  -- definition of S is OK
+
+  p :: Proxy (MkS True False)  -- this is rejected
+  p = Proxy
+
+This proposal does *not* affect normal GADT constructors. So this still works::
+
+  data T a where
+    MkT :: Bool -> T Bool
+
+  p :: Proxy (MkT True)
+  p = Proxy
+
+Effect and Interactions
+-----------------------
+
+* Much code can be deleted. A partial list is on the `ticket tracking the implemenration of this proposal <https://gitlab.haskell.org/ghc/ghc/-/issues/22298#checklist-of-things-to-remove>`_.
+
+* The inconsistency in the Motivation_ section is gone.
+
+Costs and Drawbacks
+-------------------
+
+* Putting a constraint on a kind is very modestly useful, and this proposal destroys that use case. Instead, users are welcome to constrain data constructors instead of the type itself.
+
+* If and when we get full dependent types -- including the ability to have type class constraints in kinds (which is not permitted today) -- we may need to reintroduce
+this feature, thus making the work of removing it redundant. However, it is quite likely that the new implementation will be much more well-grounded than the
+current one. It is even possible that it will be easier to implement this feature correctly after the current incorrect implementation is ripped out.
```

Comment:
Broken list formatting:

```suggestion
* If and when we get full dependent types -- including the ability to have type class constraints in kinds (which is not permitted today) -- we may need to reintroduce
  this feature, thus making the work of removing it redundant. However, it is quite likely that the new implementation will be much more well-grounded than the
  current one. It is even possible that it will be easier to implement this feature correctly after the current incorrect implementation is ripped out.
```